### PR TITLE
Add support for fuzz testing

### LINKED ISSUE #19: Fuzzy test gfx-memory
Blocked on https://github.com/gfx-rs/gfx/issues/3278
We should have tests that just spam a whole ton of allocations with various randomized memory requirements.

### PR DESCRIPTION
**Pull request by GabrielMajeri**
Uses `cargo-fuzz` to run fuzz tests.

Fuzzes the dedicated, general, and linear allocators.

Closes #19

### CODE REVIEW

File: gfx-memory/fuzz/fuzz_targets/dedicated.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#![no_main]
+
+use gfx_backend_empty::Backend;
+use gfx_memory::{Allocator, DedicatedAllocator, DedicatedBlock};
+use gfx_memory_fuzz::{create_device, Allocation};
+use hal::{memory::Properties as MemoryProperties, MemoryTypeId};
+use libfuzzer_sys::fuzz_target;
+
+fuzz_target!(|allocations: Vec<Allocation>| {
```
Comment (GabrielMajeri) [inline gfx-memory/fuzz/fuzz_targets/dedicated.rs]: The fuzzing strategy currently is to generate a random number of allocations of random size and alignment, allocate them in order, then deallocate them all

File: gfx-memory/fuzz/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+libfuzzer-sys = "0.3"
+arbitrary = { version = "0.4", features = ["derive"] }
+gfx-memory = { path = ".." }
+hal = { package = "gfx-hal", version = "0.5" }
+gfx-backend-empty = { git = "https://github.com/GabrielMajeri/gfx", branch = "hal-0.5" }
+
+[patch.crates-io]
+gfx-hal = { git = "https://github.com/GabrielMajeri/gfx", branch = "hal-0.5" }
+
+# Prevent this from interfering with workspaces
```
Comment (kvark) [inline gfx-memory/fuzz/Cargo.toml]: why shouldn't it be in the project workspace?

**Review by kvark — CHANGES_REQUESTED** [review]
Amazing work, thank you for integrating cargo-fuzz!
At the high level, is there a benefit from having fuzz sub-projects in each of the crates we have?

Sub-question (1): does it have to be a sub-project instead of just some dev-dependencies and code gated behind features?

Sub-question (2): would it be simpler to just have one top-level project in the repository that depends on both gfx-memory and gfx-descriptor and does the fuzzing?

**Comment by GabrielMajeri** [comment]
@kvark Most of the `fuzz` subdirectory was generated by `cargo fuzz init`, as [described in the tutorial](https://rust-fuzz.github.io/book/cargo-fuzz/tutorial.html). I'm not sure why they prefer a separate project just for fuzzing (I think it might be related to the fact they currently require nightly Rust for `libfuzzer`).

It should be possible to keep the configuration and dependencies behind feature flags, but we'd still need to add the following to the package's `Cargo.toml`:
```toml
[package.metadata]
cargo-fuzz = true
```
and I'm not sure what the consequences would be.

On the other hand, I can move the fuzz project to the top-level directory. It's in the `gfx-memory` subdir because I was only experimenting with this crate for now.

File: gfx-memory/fuzz/Cargo.toml (hunk 2 above)
Comment (GabrielMajeri) [reply gfx-memory/fuzz/Cargo.toml]: Seems it can cause issues: https://github.com/rust-fuzz/cargo-fuzz/issues/80

**Comment by kvark** [comment]
>On the other hand, I can move the fuzz project to the top-level directory.

yep, let's do that! One fuzz project in the workspace to fuzz them all :)

File: fuzz/fuzz_targets/general.rs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    fn arbitrary(u: &mut arbitrary::Unstructured) -> Result<Self, arbitrary::Error> {
+        let config = GeneralConfig {
+            block_size_granularity: *u.choose(POWERS_OF_TWO)?,
+            max_chunk_size: *u.choose(POWERS_OF_TWO)?,
+            min_device_allocation: *u.choose(POWERS_OF_TWO)?,
+        };
+        // Need to make sure the values make some sense
+        if config.min_device_allocation > config.max_chunk_size {
+            return Err(arbitrary::Error::IncorrectFormat);
+        }
```
Comment (GabrielMajeri) [inline fuzz/fuzz_targets/general.rs]: I'm not sure about these lines.

If I don't constraint the values to be powers of two, then 99.9% of the time the fuzzer will generate random values, trigger an assertion, and then panic. But I think these are not "interesting" crashes.

File: fuzz/fuzz_targets/general.rs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            config,
+            allocations,
+        };
+        Ok(input)
+    }
+}
+
+libfuzzer_sys::fuzz_target!(|input: FuzzingInput| {
+    let allocator =
+        GeneralAllocator::new(MemoryTypeId(0), Properties::DEVICE_LOCAL, input.config, 1);
```
Comment (GabrielMajeri) [inline fuzz/fuzz_targets/general.rs]: Also, is it ok to hardcode `DEVICE_LOCAL`, or does it change the allocation behavior a lot and should be randomly selected?

File: fuzz/fuzz_targets/general.rs (hunk 3 above)
Comment (kvark) [reply fuzz/fuzz_targets/general.rs]: Only the `block_size_granularity` should be a power of two. Could you try leaving the other ones non-power-of-two?

File: fuzz/fuzz_targets/general.rs (hunk 4 above)
Comment (kvark) [reply fuzz/fuzz_targets/general.rs]: let's just pass `Properties::empty()` there, for now

File: fuzz/fuzz_targets/general.rs (hunk 3 above)
Comment (GabrielMajeri) [reply fuzz/fuzz_targets/general.rs]: I've tried setting it to something arbitrary and got
```
thread '<unnamed>' panicked at 'Max chunk size must be power of two',
```
This seems to come from an assertion in `GeneralAllocator::new`. Is this not intended?

**Comment by GabrielMajeri** [comment]
That should be all of the allocators. I can squash the commits if needed.

File: fuzz/fuzz_targets/general.rs (hunk 3 above)
Comment (kvark) [reply fuzz/fuzz_targets/general.rs]: sorry, I missed that one

**Review by kvark — APPROVED** [review]
Very nice, thank you!
I'm surprised nothing showed up during your tests:)

**Pull request merged**